### PR TITLE
Add elevator system

### PR DESCRIPTION
**Pull request by nlamprian**
Compared to the classic gazebo elevator implementation:

* the system has an equivalent interface - int command to send the elevator to a given floor
* the model has doors attached to the floors (not the cabin), and the system manages them appropriately
* more features in both the model and system that aim to more closely approximate a real elevator

Tested in Dome. Let me know to rebase if you think it will work in Citadel.

### CODE REVIEW

**Comment by nkoenig** [comment]
Thank you for the contribution. Overall this PR looks great. The only problem is that we would like to avoid the inclusion of `boost` due to dependency and cross-platform issues. 

I've found [afsm](https://github.com/zmij/afsm) which claims that migration from `boost::msm` is easy. Do you think you could put `afsm` in a `vender` sub-directory in your plugin and swap out `boost` for `afsm`?

**Comment by nlamprian** [comment]
Sure, I understand. Looks like an easy fix, I'll give it a try.

Just to be clear, you are suggesting that I copy the repo in the `vender` subdirectory, right? (and not configure it, for example, as a cmake external project or a git submodule)

**Comment by nkoenig** [comment]
> Just to be clear, you are suggesting that I copy the repo in the `vender` subdirectory, right? (and not configure it, for example, as a cmake external project or a git submodule)

Yeah, that's correct.

**Comment by chapulina** [comment]
> Looks like an easy fix, I'll give it a try.

Hi @nlamprian , are you planning to follow up on this PR? Thanks!

**Comment by nlamprian** [comment]
Yes, I just need to find the time. Last time I tried I got stuck trying to configure afsm in the project. The change is not trivial after all.

**Comment by chapulina** [comment]
@nlamprian , how about moving this plugin to the `examples` folder? The requirements for plugins there are looser than the official ones because we don't package and ship them. I think it would be ok to keep boost as a dependency if the plugin was there.

**Comment by nlamprian** [comment]
It's been that long, huh? :grinning: 

I don't mind. Either we go with the example and I do it now, or we stay with the system and I do it within the month.

**Comment by chapulina** [comment]
> Either we go with the example and I do it now, or we stay with the system and I do it within the month.

I'll leave that up to you :wink: Thanks!

**Comment by chapulina** [comment]
Hi @nlamprian , any updates? Let us know how we can help. Thanks!

**Comment by matosinho** [comment]
@nlamprian I'm interested in this plugin as well, so if there is a way for me to contribute, let me know.

**Comment by nlamprian** [comment]
Sorry for the broken promises and for keeping this on hold all this time.

Back in August, after I upgraded my setup, the system wasn't working as intended. The lidars were keeping the doors open. Their data didn't make sense and I couldn't figure out what was wrong soon enough.

Regardless, I upgraded again and things seem to be in order. So, I'll get right to it!

Btw I trying to visualize the lidars and gazebo segfaults in VisualizeLidar::Update with "Address not mapped to object". I copied the configuration from the respective example (which works). Any idea what might be wrong?

**Comment by chapulina** [comment]
> Sorry for the broken promises and for keeping this on hold all this time.

No worries, we're all busy :wink: 

> Btw I trying to visualize the lidars and gazebo segfaults in VisualizeLidar::Update with "Address not mapped to object".

I'm not able to reproduce using `gpu_lidar_sensor.sdf`. Would you mind opening an issue with steps to reproduce? Thanks!

**Comment by nlamprian** [comment]
I'm done with the migration, but the system doesn't work. I'll need to do some debugging.

I pushed a commit to get your opinion. I added an argument in `gz_add_system` for specifying target includes. This is necessary to pickup the headers under the `vender` sub-directory.

I opened an issue for the bug in #1202

**Comment by nlamprian** [comment]
Migration is complete. I hope I made it worth the wait.

The dependencies (afsm, metapushkin) are under the vender directory as requested. I grabbed the headers and license out of the repos.

Everything else stays the same.

File: src/systems/elevator/utils/DoorTimer.hh
Comment (chapulina) [inline src/systems/elevator/utils/DoorTimer.hh:59]: I see you using `chrono` types in some places, but `double` in others. It would be nice to use `chrono` everywhere for consistency and clarity.

File: src/systems/elevator/utils/DoorTimer.cc
Comment (chapulina) [inline src/systems/elevator/utils/DoorTimer.cc:81]: Is it ok just not to pass the ECM, since it's not used?

File: src/systems/elevator/utils/DoorTimer.cc
Comment (chapulina) [inline src/systems/elevator/utils/DoorTimer.cc:48]: This information doesn't seem to be used outside of `Update`, do we need to keep it as a member variable?

File: src/systems/elevator/utils/DoorTimer.hh
Comment (chapulina) [inline src/systems/elevator/utils/DoorTimer.hh:57]: It's good to document that `Configure` needs to be called after this, otherwise the timeout won't be updated.

A higher level recommendation though would be to get rid of the `SetWaitDuration` and `Configure` functions and instead do it all in the constructor to follow the RAII idiom. The duration can have a default value so it's optional. This way, there's no way the user can call functions in the wrong order or forget to call them.

File: src/systems/elevator/vender/metapushkin/LICENSE
Comment (chapulina) [inline src/systems/elevator/vender/metapushkin/LICENSE:1]: @j-rivero , do you see any issues mixing this license here?

File: test/worlds/elevator.sdf
Comment (chapulina) [inline test/worlds/elevator.sdf:7]:
```suggestion
      <real_time_factor>0</real_time_factor>
```

To run the test as fast as possible

File: test/worlds/elevator.sdf
Comment (chapulina) [inline test/worlds/elevator.sdf:21]: I think `Sensors`, `SceneBroadcaster`, the light and maybe even the ground plane aren't needed for the test

File: src/systems/elevator/Elevator.hh
Comment (chapulina) [inline src/systems/elevator/Elevator.hh:125]: Can this be a `unique_ptr`?

File: src/systems/elevator/Elevator.cc
Comment (chapulina) [inline src/systems/elevator/Elevator.cc:180]: Should we just print an error if the period is <= 0 and stop loading the plugin, or does the plugin have any functionality with zero?

File: src/systems/elevator/Elevator.cc
Comment (chapulina) [inline src/systems/elevator/Elevator.cc:202]:
```suggestion
  if (!this->dataPtr->InitDoors(doorJointPrefix, topicPrefix, _ecm))
      return;
```

style nit

File: src/systems/elevator/state_machine/ActionsImpl.hh
Comment (chapulina) [inline src/systems/elevator/state_machine/ActionsImpl.hh:51]:
```suggestion
    if (trigger)
      _fsm.process_event(events::NewTarget());
```

File: src/systems/elevator/state_machine/ActionsImpl.hh
Comment (chapulina) [inline src/systems/elevator/state_machine/ActionsImpl.hh:67]:
```suggestion
    if (data->targets.front() == data->system->state)
      data->targets.pop_front();
```

File: src/systems/elevator/state_machine/ActionsImpl.hh
Comment (chapulina) [inline src/systems/elevator/state_machine/ActionsImpl.hh:60]: Missing some docs

File: src/systems/elevator/state_machine/ElevatorStateMachineImpl.hh
Comment (chapulina) [inline src/systems/elevator/state_machine/ElevatorStateMachineImpl.hh:129]: I think this defeats the purpose of using the PIMPL pattern. If other classes need to use this data, then it shouldn't be private, or there should be setters and getters for it.

File: src/systems/elevator/state_machine/StatesImpl.hh
Comment (chapulina) [inline src/systems/elevator/state_machine/StatesImpl.hh:145]: Is there a reason to override the tolerance here? Maybe this should be the default value?

File: src/systems/elevator/state_machine/StatesImpl.hh
Comment (chapulina) [inline src/systems/elevator/state_machine/StatesImpl.hh:186]:
```suggestion
  private: void OnTimeout()
      {
        this->triggerEvent(); 
      }
```

**Review by chapulina — CHANGES_REQUESTED** [review]
Thanks again for the PR, it has a lot of functionality! Unfortunately the big ones take a lot of effort to review, so we're going slowly. I did a first high-level review now.

We're preparing the final `ign-gazebo4` releases now, since Dome will EOL in a few days. I think we won't have time to finish iterating on this PR by then :confused: So it will need to be retargeted at another branch. You asked above if this could work with Citadel, and I think that should be reasonable, but I'll leave it up to whatever is easier for you.

Thanks again!

**Comment by nlamprian** [comment]
I backported the branch to citadel and tested it successfully. Now, how do you want this done? New PR / How do we proceed with the review?

**Comment by chapulina** [comment]
> I backported the branch to citadel and tested it successfully. Now, how do you want this done? New PR / How do we proceed with the review?

Thanks! We can keep this PR, you just need to rebase onto `ign-gazebo3` and change the PR base branch.

File: src/systems/elevator/utils/DoorTimer.cc
Comment (nlamprian) [reply src/systems/elevator/utils/DoorTimer.cc:48]: It's used to detect transitions between iterations, and I cannot set it as static in the method as there is more than one instance. Similarly, isActive goes into the same basket.

Comment (nlamprian) [reply src/systems/elevator/utils/DoorTimer.cc:48]: It's used in Configure. It's re-initialized when the timer is started.

File: test/worlds/elevator.sdf
Comment (nlamprian) [reply test/worlds/elevator.sdf:7]: Hmm, never crossed my mind :+1:

File: src/systems/elevator/utils/DoorTimer.hh
Comment (nlamprian) [reply src/systems/elevator/utils/DoorTimer.hh:57]: I agree in regards to SetWaitDuration. I did it like this because the duration doesn't exist at the moment of creation. Not sure if I can consider it optional and give it a default value, so I think I'll turn the timer variable into a unique_ptr to get around that.

For Configure though, it's called every time the timer is started, so it needs to stay. Does it make you draw a correlation to ISystemConfigure by giving it this name? The same concept applies to the joint monitor.

File: test/worlds/elevator.sdf
Comment (nlamprian) [reply test/worlds/elevator.sdf:21]: Done. On a side note, the test passes, but now, in citadel, I get this error

```
[Dbg] [SimulationRunner.cc:422] Creating PostUpdate worker threads: 2
[Dbg] [SimulationRunner.cc:435] Creating postupdate worker thread (0)
[Err] [JointPositionController.cc:214] [JointPositionController]: Detected an invalid <joint_index> parameter. The index specified is [0] but the joint only has [0] index[es]. This controller will be ignored
```

File: src/systems/elevator/Elevator.cc
Comment (nlamprian) [reply src/systems/elevator/Elevator.cc:180]: Zero means no throttling. I'll add it to the documentation (or treat it as an error as you suggested)

File: src/systems/elevator/state_machine/StatesImpl.hh
Comment (nlamprian) [reply src/systems/elevator/state_machine/StatesImpl.hh:145]: I think I have the open state at 5cm because it takes some "wasted" time to reach steady-state, and the close state at 2cm because, otherwise, I was seeing the cabin move before the door was "fully" closed. I can probably live with both at 2cm.

File: src/systems/elevator/Elevator.hh
Comment (nlamprian) [reply src/systems/elevator/Elevator.hh:125]: It's shared with the state machine

File: src/systems/elevator/state_machine/ElevatorStateMachineImpl.hh
Comment (nlamprian) [reply src/systems/elevator/state_machine/ElevatorStateMachineImpl.hh:129]: I'll make it public. In this case, I didn't have PIMPL in mind. I used dataPtr so I could pass data to the actions/guards/states.

Beyond that, I used PIMPL in all classes because the documentation says so. But does it serve any purpose here? All classes are internal to the elevator system, and they are compiled together in one library.

File: src/systems/elevator/vender/metapushkin/LICENSE
Comment (nkoenig) [reply src/systems/elevator/vender/metapushkin/LICENSE:1]: The Artistic License 2.0 is compatible with Apache 2.0

**Review by nkoenig — APPROVED** [review]
This works for me. I have a few style changes that I'll make in a separate PR.

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-03-01-citadel-edifice-fortress/1313/1

**Comment by osrf-triage** [comment]
This pull request has been mentioned on **Gazebo Community**. There might be relevant details there:

https://community.gazebosim.org/t/new-ignition-releases-2022-04-13-fortress-edifice/1367/1

**Pull request merged**
